File: cli/src/orchestration/executor.ts
/**
 * Concurrent agent executor for multi-agent orchestration.
 * Manages task checkout/release, concurrency limits, and lifecycle events.
 */

import { EventEmitter } from 'node:events';
import type { TaskDefinition, OrchestratorConfig } from './types.js';

export type ExecutorEvent =
  | { type: 'task:start'; taskId: string; agentSlug: string }
  | { type: 'task:complete'; taskId: string; result?: string }
  | { type: 'task:fail'; taskId: string; error: string }
  | { type: 'task:timeout'; taskId: string };

export class OrchestratorExecutor {
  private config: OrchestratorConfig;
  private tasks: Map<string, TaskDefinition> = new Map();
  private locks: Map<string, string> = new Map(); // taskId -> agentSlug
  private activeCount = 0;
  private emitter = new EventEmitter();
  private timers: Map<string, ReturnType<typeof setTimeout>> = new Map();
  private stopped = false;

  constructor(config: OrchestratorConfig) {
    this.config = config;
  }

  /** Subscribe to executor events. */
  on(handler: (event: ExecutorEvent) => void): () => void {
    this.emitter.on('event', handler);
    return () => {
      this.emitter.off('event', handler);
    };
  }

  private emit(event: ExecutorEvent): void {
    this.emitter.emit('event', event);
  }

  /**
   * Atomic task checkout -- prevents duplicate assignment.
   * Returns true if the task was successfully locked for this agent.
   */
  async checkout(taskId: string, agentSlug: string): Promise<boolean> {
    // Already locked by someone else
    if (this.locks.has(taskId)) return false;

    const task = this.tasks.get(taskId);
    if (!task) return false;
    if (task.status !== 'pending') return false;

    // Concurrency limit
    if (this.activeCount >= this.config.maxConcurrentAgents) return false;

    // Lock it
    this.locks.set(taskId, agentSlug);
    task.status = 'locked';
    task.assignedAgent = agentSlug;
    this.tasks.set(taskId, task);
    return true;
  }

  /**
   * Release a task on completion or failure.
   */
  async release(
    taskId: string,
    result?: string,
    error?: string,
  ): Promise<void> {
    const task = this.tasks.get(taskId);
    if (!task) return;

    // Clear timeout timer
    const timer = this.timers.get(taskId);
    if (timer) {
      clearTimeout(timer);
      this.timers.delete(taskId);
    }

    this.locks.delete(taskId);

    if (error) {
      task.status = 'failed';
      task.result = error;
      task.completedAt = Date.now();
      this.emit({ type: 'task:fail', taskId, error });
    } else {
      task.status = 'completed';
      task.result = result;
      task.completedAt = Date.now();
      this.emit({ type: 'task:complete', taskId, result });
    }

    this.tasks.set(taskId, task);
    if (this.activeCount > 0) this.activeCount--;
  }

  /**
   * Submit a task for execution.
   * Stores the task; actual execution happens during runBatch or by
   * the external executor loop calling checkout/release.
   */
  async submit(task: TaskDefinition): Promise<void> {
    this.tasks.set(task.id, { ...task, status: 'pending' });
  }

  /**
   * Run multiple tasks concurrently, respecting maxConcurrentAgents.
   * Assigns tasks to their assignedAgent (or marks them as in_progress).
   * Returns final task states.
   */
  async runBatch(
    tasks: TaskDefinition[],
  ): Promise<Map<string, TaskDefinition>> {
    this.stopped = false;

    // Register all tasks
    for (const t of tasks) {
      await this.submit(t);
    }

    const pending = tasks.map((t) => t.id);
    const results = new Map<string, TaskDefinition>();

    // Process in waves respecting concurrency limit
    const runNext = async (): Promise<void> => {
      while (pending.length > 0 && !this.stopped) {
        // Fill up to max concurrent
        const toStart: string[] = [];
        while (
          pending.length > 0 &&
          this.activeCount < this.config.maxConcurrentAgents
        ) {
          const taskId = pending.shift()!;
          const task = this.tasks.get(taskId);
          if (!task) continue;

          const agent = task.assignedAgent ?? 'default';
          const locked = await this.checkout(taskId, agent);
          if (locked) {
            toStart.push(taskId);
            this.activeCount++;
          } else {
            // Put back for retry
            pending.push(taskId);
            break;
          }
        }

        // Start timeout timers and mark as in_progress
        const completions = toStart.map((taskId) => {
          const task = this.tasks.get(taskId)!;
          task.status = 'in_progress';
          this.tasks.set(taskId, task);
          this.emit({
            type: 'task:start',
            taskId,
            agentSlug: task.assignedAgent ?? 'default',
          });

          // Set timeout
          return new Promise<void>((resolve) => {
            const timer = setTimeout(async () => {
              const current = this.tasks.get(taskId);
              if (current && current.status === 'in_progress') {
                this.emit({ type: 'task:timeout', taskId });
                await this.release(taskId, undefined, 'Task timed out');
              }
              resolve();
            }, this.config.taskTimeout);
            this.timers.set(taskId, timer);

            // Simulate immediate completion for batch mode
            // (real execution is driven externally)
            resolve();
          });
        });

        await Promise.all(completions);

        // Collect results for started tasks
        for (const taskId of toStart) {
          const task = this.tasks.get(taskId);
          if (task) results.set(taskId, task);
        }

        // If all tasks started, break
        if (pending.length === 0) break;
      }
    };

    await runNext();

    // Return all task states
    for (const [id, task] of this.tasks) {
      results.set(id, task);
    }
    return results;
  }

  /** Get currently active (in_progress) tasks. */
  getActive(): TaskDefinition[] {
    const active: TaskDefinition[] = [];
    for (const task of this.tasks.values()) {
      if (task.status === 'in_progress' || task.status === 'locked') {
        active.push(task);
      }
    }
    return active;
  }

  /** Stop all running tasks. */
  async stopAll(): Promise<void> {
    this.stopped = true;
    for (const [taskId, timer] of this.timers) {
      clearTimeout(timer);
      this.timers.delete(taskId);
    }
    for (const task of this.tasks.values()) {
      if (task.status === 'in_progress' || task.status === 'locked') {
        task.status = 'failed';
        task.result = 'Stopped by orchestrator';
        task.completedAt = Date.now();
        this.locks.delete(task.id);
      }
    }
    this.activeCount = 0;
  }
}

File: internal/orchestration/executor.go
package orchestration

import (
	"fmt"
	"sync"
	"time"
)

// ExecutorEvent is emitted when a task transitions state.
type ExecutorEvent struct {
	Type      string // "task:start", "task:complete", "task:fail", "task:timeout"
	TaskID    string
	AgentSlug string
	Result    string
	Error     string
}

// Executor manages task lifecycle with concurrency limits and timeout enforcement.
type Executor struct {
	config      OrchestratorConfig
	tasks       map[string]*TaskDefinition
	locks       map[string]string // taskID → agentSlug
	activeCount int
	listeners   []func(ExecutorEvent)
	timers      map[string]*time.Timer
	stopped     bool
	mu          sync.Mutex
}

// NewExecutor returns an Executor using the provided configuration.
func NewExecutor(config OrchestratorConfig) *Executor {
	return &Executor{
		config:  config,
		tasks:   make(map[string]*TaskDefinition),
		locks:   make(map[string]string),
		timers:  make(map[string]*time.Timer),
	}
}

// OnEvent registers a handler for task events. Returns an unsubscribe function.
func (e *Executor) OnEvent(handler func(ExecutorEvent)) func() {
	e.mu.Lock()
	defer e.mu.Unlock()
	e.listeners = append(e.listeners, handler)
	idx := len(e.listeners) - 1
	return func() {
		e.mu.Lock()
		defer e.mu.Unlock()
		if idx < len(e.listeners) {
			e.listeners = append(e.listeners[:idx], e.listeners[idx+1:]...)
		}
	}
}

// Submit adds a task to the executor's queue (status "pending").
// Returns an error if a task with the same ID already exists.
func (e *Executor) Submit(task TaskDefinition) error {
	e.mu.Lock()
	defer e.mu.Unlock()
	if e.stopped {
		return fmt.Errorf("executor is stopped")
	}
	if _, exists := e.tasks[task.ID]; exists {
		return fmt.Errorf("task %q already submitted", task.ID)
	}
	task.Status = "pending"
	copy := task
	e.tasks[task.ID] = &copy
	return nil
}

// Checkout atomically assigns taskID to agentSlug and starts the task.
// Returns (true, nil) on success, (false, nil) if the task is already locked
// or the concurrency limit is reached, or (false, err) on other errors.
func (e *Executor) Checkout(taskID, agentSlug string) (bool, error) {
	e.mu.Lock()
	defer e.mu.Unlock()

	if e.stopped {
		return false, fmt.Errorf("executor is stopped")
	}

	task, ok := e.tasks[taskID]
	if !ok {
		return false, fmt.Errorf("task %q not found", taskID)
	}
	if task.Status != "pending" {
		return false, nil
	}
	if _, locked := e.locks[taskID]; locked {
		return false, nil
	}
	if e.config.MaxConcurrentAgents > 0 && e.activeCount >= e.config.MaxConcurrentAgents {
		return false, nil
	}

	task.Status = "in_progress"
	task.AssignedAgent = agentSlug
	e.locks[taskID] = agentSlug
	e.activeCount++

	// Schedule timeout if configured.
	if e.config.TaskTimeout > 0 {
		timer := time.AfterFunc(e.config.TaskTimeout, func() {
			e.handleTimeout(taskID, agentSlug)
		})
		e.timers[taskID] = timer
	}

	e.emit(ExecutorEvent{Type: "task:start", TaskID: taskID, AgentSlug: agentSlug})
	return true, nil
}

// Release marks a task as completed or failed and frees the agent slot.
// Pass a non-nil result string for success, a non-nil err string for failure.
func (e *Executor) Release(taskID string, result *string, errStr *string) error {
	e.mu.Lock()
	defer e.mu.Unlock()

	task, ok := e.tasks[taskID]
	if !ok {
		return fmt.Errorf("task %q not found", taskID)
	}
	if _, locked := e.locks[taskID]; !locked {
		return fmt.Errorf("task %q is not checked out", taskID)
	}

	agentSlug := e.locks[taskID]
	delete(e.locks, taskID)
	if e.activeCount > 0 {
		e.activeCount--
	}

	// Cancel timeout timer.
	if timer, ok := e.timers[taskID]; ok {
		timer.Stop()
		delete(e.timers, taskID)
	}

	task.CompletedAt = time.Now().UnixMilli()

	if errStr != nil && *errStr != "" {
		task.Status = "failed"
		e.emit(ExecutorEvent{
			Type:      "task:fail",
			TaskID:    taskID,
			AgentSlug: agentSlug,
			Error:     *errStr,
		})
	} else {
		task.Status = "completed"
		if result != nil {
			task.Result = *result
		}
		e.emit(ExecutorEvent{
			Type:      "task:complete",
			TaskID:    taskID,
			AgentSlug: agentSlug,
			Result:    task.Result,
		})
	}
	return nil
}

// GetActive returns a snapshot of all currently in-progress tasks.
func (e *Executor) GetActive() []TaskDefinition {
	e.mu.Lock()
	defer e.mu.Unlock()
	var out []TaskDefinition
	for _, t := range e.tasks {
		if t.Status == "in_progress" {
			out = append(out, *t)
		}
	}
	return out
}

// StopAll cancels all in-progress tasks and marks them as failed.
func (e *Executor) StopAll() error {
	e.mu.Lock()
	defer e.mu.Unlock()

	e.stopped = true

	for id, timer := range e.timers {
		timer.Stop()
		delete(e.timers, id)
	}

	for taskID, agentSlug := range e.locks {
		task := e.tasks[taskID]
		task.Status = "failed"
		task.CompletedAt = time.Now().UnixMilli()
		delete(e.locks, taskID)
		if e.activeCount > 0 {
			e.activeCount--
		}
		e.emit(ExecutorEvent{
			Type:      "task:fail",
			TaskID:    taskID,
			AgentSlug: agentSlug,
			Error:     "executor stopped",
		})
	}
	return nil
}

// handleTimeout is called by the per-task timer; fires a timeout event.
func (e *Executor) handleTimeout(taskID, agentSlug string) {
	e.mu.Lock()
	defer e.mu.Unlock()

	task, ok := e.tasks[taskID]
	if !ok {
		return
	}
	if task.Status != "in_progress" {
		return
	}

	task.Status = "failed"
	task.CompletedAt = time.Now().UnixMilli()
	delete(e.locks, taskID)
	delete(e.timers, taskID)
	if e.activeCount > 0 {
		e.activeCount--
	}

	e.emit(ExecutorEvent{
		Type:      "task:timeout",
		TaskID:    taskID,
		AgentSlug: agentSlug,
		Error:     "task timed out",
	})
}

// emit fires all registered listeners. Must be called with mu held.
func (e *Executor) emit(event ExecutorEvent) {
	for _, fn := range e.listeners {
		go fn(event) // fire async to avoid deadlock with mu held
	}
}
